runtime/io.c

```diff
@@ -69,15 +69,17 @@ CAMLexport struct channel * caml_all_opened_channels = NULL;
 
 /* Functions shared between input and output */
 
-CAMLexport struct channel * caml_open_descriptor_in(int fd)
+CAMLexport struct channel * caml_open_descriptor_in_exn(int fd, value * exn)
 {
   struct channel * channel;
 
   channel = (struct channel *) caml_stat_alloc(sizeof(struct channel));
   channel->fd = fd;
-  caml_enter_blocking_section();
+  *exn = caml_enter_blocking_section_exn();
```

Comment:
There are two possible conventions:
- Only write to *exn if there is an exception.
- Always write to *exn; i.e. treat exn as an output parameter.
In the latter case, the caller does not need to initialise *exn. Question for seasoned C programmers: are there any coding guidelines saying that one is better or more expected than the other?

Reply:
IMHO, a common pattern for out parameters is to not write them when there is no corresponding result, and signal by another channel (i.e., the return value of the function) that the out parameter has been (or not) written.

Reply:
Ok but then I would think that exn is the error code, and the channel is the output param. I.e. the signature should be
```c
CAMLexport value caml_open_descriptor_in_exn(int fd, struct channel ** chan)
```
instead of:
```c
CAMLexport struct channel * caml_open_descriptor_in_exn(int fd, value * exn)
```
I thought about doing this, but preferred a change that would be perceived as smaller. It seems that we will have to do it anyway.

Reply:
I'd rather not change the return type of the function. I think this is misleading. Why cannot you guarantee that [exn] is written iff the return value is NULL?

Reply:
I could, but in addition I would prefer if there is a convention that for `_exn` functions one has to first check for `Is_exception_result`.

Reply:
Sure. But this is a convention to be followed by the caller side. This does not prevent you to follow the usual convention that out parameters are only written in case it is needed. Of course, there will be some redundancy, but is that an issue?